delta/plugins/storage/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/storages/operations/s3/S3StorageSaveFile.scala

```diff
@@ -1,76 +1,166 @@
-                      uuid = uuid,
-                      bytes = bytes,
-                      digest = digest,
-                      origin = Client,
-                      location = s3Result.location.withPath(Slash(path)),
-                      path = Uri.Path(key)
-                    )
-                  )
-              })
-            case Some(_) => Future.failed(fileAlreadyExistException)
-          }
+    val key                        = path.toString()
+    val fileData: Stream[IO, Byte] = convertStream(entity.dataBytes)
+
+    (for {
+      _          <- log(key, s"Checking for object existence")
+      _          <- validateObjectDoesNotExist(key)
+      _          <- log(key, s"Beginning multipart upload")
+      maybeEtags <- uploadFileMultipart(fileData, key)
+      _          <- log(key, s"Finished multipart upload. Etag by part: $maybeEtags")
+      attr       <- collectFileMetadata(fileData, key, uuid, maybeEtags)
+    } yield attr)
+      .adaptError { err => UnexpectedSaveError(key, err.getMessage) }
+  }
+
+  private def validateObjectDoesNotExist(key: String) =
+    getFileAttributes(key).redeemWith(
+      {
+        case _: NoSuchKeyException => IO.unit
+        case e                     => IO.raiseError(e)
+      },
+      _ => IO.raiseError(ResourceAlreadyExists(key))
+    )
+
+  private def convertStream(source: Source[ByteString, Any]): Stream[IO, Byte] =
+    StreamConverter(
+      source
+        .flatMapConcat(x => Source.fromIterator(() => x.iterator))
+        .mapMaterializedValue(_ => NotUsed)
+    )
+
+  private def uploadFileMultipart(fileData: Stream[IO, Byte], key: String): IO[List[Option[ETag]]] =
+    fileData
+      .through(
+        s3.uploadFileMultipart(
+          bucket,
+          FileKey(NonEmptyString.unsafeFrom(key)),
+          partSizeMB,
+          uploadEmptyFiles = true,
+          multipartETagValidation = multipartETagValidation.some
+        )
+      )
+      .compile
+      .to(List)
+
+  private def getFileAttributes(key: String): IO[GetObjectAttributesResponse] =
+    client
+      .getObjectAttributes(
+        GetObjectAttributesRequest
+          .builder()
+          .bucket(bucket.value.value)
+          .key(key)
+          .objectAttributes(ObjectAttributes.OBJECT_SIZE) // TODO get all values
+          .build()
       )
-    ).adaptError {
-      case `fileAlreadyExistException` => ResourceAlreadyExists(key)
-      case err: S3Exception            => UnexpectedSaveError(key, err.toString())
-      case err                         => UnexpectedSaveError(key, err.getMessage)
+
+  private def collectFileMetadata(
+      bytes: Stream[IO, Byte],
+      key: String,
+      uuid: UUID,
+      maybeEtags: List[Option[ETag]]
+  ): IO[FileStorageMetadata] =
+    maybeEtags.sequence match {
+      case Some(onlyPartETag :: Nil) =>
+        // TODO our tests expect specific values for digests and the only algorithm currently used is SHA-256.
```

Comment:
Can be done separately, and will be much simpler if we just use a regular put object (as @imsdu suggested)